x/icacallbacks/module_ibc.go

```diff
@@ -93,30 +94,70 @@ func (im IBCModule) NegotiateAppVersion(
 	return proposedVersion, nil
 }
 
-// GetTxMsgData returns the msgs from an ICA transaction and can be reused across authentication modules
-func GetTxMsgData(ctx sdk.Context, ack channeltypes.Acknowledgement, logger log.Logger) (*types.ICATxResponse, error) {
+// UnpackAcknowledgementResponse returns the msgs from an ICA transaction and can be reused across authentication modules
+func UnpackAcknowledgementResponse(ctx sdk.Context, ack []byte, logger log.Logger) (*types.ICATxResponse, error) {
+	// NOTE: Is this safe to assume? According to the ics-27 spec, errors on host chains should be reflected in the acknowledgement type
+	// so Acknowledgement_Result should only be returned if the transaction was successful
+	// However, I recall seeing a case where the acknowledgement type was Acknowledgement_Result but the transaction failed
+	//
+	// -> If my memory's correct, I think a failure on the host could come back as either an AcknowledgementError
+	//    or it could be an AcknowledgementResult with data length 0
+	//    It's a little concering that the switch statement now just assumes the data length 0 means it's sdk 46
+	//
+	// It's also unclear to me whether this timeout case below is possible
+	// In any case, we should just test this thoughly with the following cases:
+	//   Success
```

Comment:
I don't think this sequence is possible, because the host chain should process either a timeout or an error, but not both

Reply:
We identify it with a nil ack out of convenience here https://github.com/notional-labs/stride/blob/acd8c57188652ce42c53a09c82a2d5bbe49b83af/x/stakeibc/module_ibc.go#L174

The timeout actually flows down a different code path (`OnTimeout`), so it probably makes sense to stop using a nil ack to identify timeouts and instead pass in an `ICATxResponse` with a timeout status. Sorry, should have caught this in the review - I'll push this up now.

Reply:
Ah that's right, good call out!